File: 1036-rotting-oranges/rotting-oranges.py
class Solution:
    def orangesRotting(self, grid: List[List[int]]) -> int:
        rows, cols = len(grid), len(grid[0])

        queue = deque()
        fresh = 0

        # Step 1: Initialize queue with all rotten oranges and count fresh ones
        for r in range(rows):
            for c in range(cols):
                if grid[r][c] == 2:
                    queue.append((r, c, 0))  # (row, col, time)
                elif grid[r][c] == 1:
                    fresh += 1

        directions = [(-1, 0), (1, 0), (0, -1), (0, 1)]  # up, down, left, right
        time_elapsed = 0

        # Step 2: BFS
        while queue:
            r, c, time = queue.popleft()
            time_elapsed = max(time_elapsed, time)

            for dr, dc in directions:
                nr, nc = r + dr, c + dc
                # Check boundaries and if it's a fresh orange
                if 0 <= nr < rows and 0 <= nc < cols and grid[nr][nc] == 1:
                    grid[nr][nc] = 2  # Mark as rotten
                    fresh -= 1
                    queue.append((nr, nc, time + 1))

        return time_elapsed if fresh == 0 else -1

File: 1036-rotting-oranges/rotting-oranges.cpp
class Solution {
public:
    int dx[4]={-1,0,1,0};
    int dy[4]={0,1,0,-1}; 
    int orangesRotting(vector<vector<int>>& grid) {
        int r=grid.size();
        int c=grid[0].size();
        int ans=0;
        queue<pair<int, int>>pq;
        for(int i=0;i<r;i++){
            for(int j=0;j<c;j++){
                if(grid[i][j]==2){
                    pq.push({i,j});
                }
            }
        }
        while(!pq.empty()){
            int sz=pq.size();
            int temp=0;
            while(sz--){
                auto p=pq.front();
                pq.pop();
                for(int k=0;k<4;k++){
                    int i=p.first+dx[k], j=p.second+dy[k];
                    if(i>=0 && j>=0 && i<r && j<c && grid[i][j]==1){
                        grid[i][j]=2;
                        temp=1;
                        pq.push({i,j});
                    }
                }
            }
            ans+=temp;
        }
        for(auto v:grid){
            for(int x:v) if(x==1) return -1;

        }
        return ans;
    }
};